### PR TITLE
nltkModels: init punkt at 3.8.1

### PR DESCRIPTION
**Pull request by happysalada**
## Description of changes

@natsukium I'm mainly looking for feedback on this PR. I think having the nltk models as packages would be useful.
Here are the questions I still have.
- those aren't python-modules, so wondering where to put them really.
- the urls don't have version, I'm not sure those models are even versioned. I've added nltk.version in there but I don't like it.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by natsukium** [comment]
Sounds good.

> those aren't python-modules, so wondering where to put them really.

Perhaps it would be better to place it to `pkgs/data/machine-learning/`, `pkgs/data/misc/` or `pkgs/data/text/`.
There is another example of mecab, `pkgs/tools/text/`.

> the urls don't have version, I'm not sure those models are even versioned. I've added nltk.version in there but I don't like it.

I found this repo. How about packaging it as `nltk-data`?
https://github.com/nltk/nltk_data

Arch and conda are also packaging it as `nltk-data`.
https://repology.org/project/nltk-data/versions
https://github.com/conda-forge/nltk_data-feedstock

**Comment by happysalada** [comment]
All those are great ideas thanks.
Let me know what you think.
If that works for you, I'll add another package that I need for unstructured api and we can merge this for other people to test and comment (if that works for you).

**Comment by happysalada** [comment]
I've added another one. We can probably abstract those two packages into one function accepting name and location directory.

**Comment by happysalada** [comment]
I've refactored things to use a single file.
I'm not 100% sure that the scopes work fine, but it's looking good.

File: pkgs/tools/text/nltk_data/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{ lib, newScope, fetchFromGitHub, unzip, stdenv }:
+let
+  base = {
+    version = "unstable-2023-02-02";
+    src = fetchFromGitHub {
+      owner = "nltk";
+      repo = "nltk_data";
+      rev = "5db857e6f7df11eabb5e5665836db9ec8df07e28";
+      sha256 = "sha256-RFywUkesoUL8tbiKLs6Pa+NpoEOq9eu1dP83n5XkNpY=";
```
Comment (natsukium) [inline pkgs/tools/text/nltk_data/default.nix:9]:
```suggestion
      hash = "sha256-RFywUkesoUL8tbiKLs6Pa+NpoEOq9eu1dP83n5XkNpY=";
```

File: pkgs/tools/text/nltk_data/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    };
+    nativeBuildInputs = [ unzip ];
+    dontBuild = true;
+    meta = with lib; {
+      description = "NLTK Data";
+      homepage = "https://github.com/nltk/nltk_data";
+      license = licenses.asl20;
+      platforms = platforms.unix;
```
Comment (natsukium) [inline pkgs/tools/text/nltk_data/default.nix:17]:
```suggestion
      platforms = platforms.all;
```
It's just data, and I don't think it needs to be limited to unix.

File: pkgs/tools/text/nltk_data/default.nix [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      maintainers = with maintainers; [ happysalada ];
+    };
+  };
+  makeNltkDataPackage = {pname, location}:
+    stdenv.mkDerivation (base // {
```
Comment (natsukium) [inline pkgs/tools/text/nltk_data/default.nix:22]:
```suggestion
    stdenvNoCC.mkDerivation (base // {
```

**Review by natsukium — COMMENTED** [review]
Awesome.
There are a limited number of packages, so I agree with combining them into one file.

I think we often only want part of the package.
Can we do a sparse checkout?

**Comment by happysalada** [comment]
All of the comments were high quality, thanks a lot!
I think I've addressed everything, let me know if I forgot anything.

File: pkgs/tools/text/nltk_data/default.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      };
+    in
+    stdenvNoCC.mkDerivation (base // {
+      inherit pname src;
+      version = base.version;
+      installPhase = ''
+        mkdir -p $out
+        unzip ${src}/${location}/${pname}.zip
+        cp -R ${pname}/ $out/
+      '';
```
Comment (natsukium) [inline pkgs/tools/text/nltk_data/default.nix:32]:
```suggestion
      installPhase = ''
        runHook preInstall

        mkdir -p $out
        unzip ${src}/${location}/${pname}.zip
        cp -R ${pname}/ $out/
        
        runHook postInstall
      '';
```

**Review by natsukium — COMMENTED** [review]
I forgot that we have to manage each hash individually.
It would be nice to update the latest version, rev, and hashes using updateScript, although it is not required.

**Comment by happysalada** [comment]
I hear you regarding the updating script. I'll tell you what, if we end up updating this often, I'll take the time to make a script.
My understanding is that it might not get that many updates, but I could be wrong.

**Review by natsukium — APPROVED** [review]

**Pull request merged**